IO/BeagleBone/CANBusBBB.cs

```diff
@@ -152,15 +236,29 @@ internal CANBusBBB(string CanName)
 		/// <param name="Data">Payload of CAN Frame. Must be at most 8 bytes.</param>
 		public void Write(uint ID, byte[] Data)
 		{
-			if (Data.Length > 8) { throw new Exception("CAN Data Length must be no more than 8 bytes"); }
+			if (!Extended && Data.Length > 8) { throw new Exception("CAN Data Length must be no more than 8 bytes for non-Extended frames"); }
+			else if (Data.Length > 64) { throw new Exception("CAN Data Length must be no more than 64 bytes for Extended frames."); }
+			int BytesWritten;
 			unsafe
 			{
-				CANFrame Frame = new CANFrame();
-				Frame.CANID = ID;
-				Frame.DataLength = (byte)Data.Length;
-				for (int i = 0; i < Data.Length; i++) { Frame.Data[i] = Data[i]; }
-				write(Socket, ref Frame, Marshal.SizeOf(Frame));
+				if (Extended)
+				{
+					ExtendedCANFrame Frame = new ExtendedCANFrame();
+					Frame.CANID = ID | CAN_EFF_FLAG;
+					Frame.DataLength = DLCToLen(LenToDLC((byte)Data.Length));
```

Comment:
Those methods are really not well named, here it looks like you are doing something then undoing it.